### PR TITLE
PUB-1254 Media User Emails

### PR DESCRIPTION
**Pull request by junaidiqbalmoj**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/PUB-1254

### Change description ###

Send email to media users on account creation.

**Does this PR introduce a breaking change?** (check one with "x")
[] Yes
[x] No

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java [hunk 1]
```diff
@@ -128,9 +134,30 @@ public Map<CreationEnum, List<?>> addUsers(List<PiUser> users, String issuerEmai
...
+                String emailSent = publicationService.sendNotificationEmailForDuplicateMediaAccount(user.getEmail(),
+                                                                            application.get().getFullName());
+                if (emailSent != null && !emailSent.isEmpty()) {
+                    ErroredPiUser erroredUser = new ErroredPiUser(user);
+                    erroredUser.setErrorMessages(Arrays.asList(emailSent));
+                    erroredAccounts.add(erroredUser);
+                }
+
+                continue;
+            }
```
Comment (kurt-uk) [inline src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java:151]: Should this not be part of the validation?

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java [hunk 2]
```diff
@@ -128,9 +134,30 @@ public Map<CreationEnum, List<?>> addUsers(List<PiUser> users, String issuerEmai
...
             }
 
+            Optional<MediaAndLegalApplication> application =
+                mediaLegalApplicationRepository.findByEmail(user.getEmail());
+
+            if (userRepository.findByEmail(user.getEmail()).isPresent()
+                && application.isPresent()) {
+                String emailSent = publicationService.sendNotificationEmailForDuplicateMediaAccount(user.getEmail(),
+                                                                            application.get().getFullName());
+                if (emailSent != null && !emailSent.isEmpty()) {
```
Comment (kurt-uk) [inline src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java:144]: this logic doesnt line up with what youve coded, this will never hit

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java [hunk 3]
```diff
@@ -128,9 +134,30 @@ public Map<CreationEnum, List<?>> addUsers(List<PiUser> users, String issuerEmai
...
+
+                continue;
+            }
+
             PiUser addedUser = userRepository.save(user);
             createdAccounts.add(addedUser.getUserId());
 
+            if (application.isPresent()) {
+                publicationService.sendNotificationEmailForSetupMediaAccount(user.getEmail(),
+                                                                             application.get().getFullName());
```
Comment (kurt-uk) [inline src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java:158]: do you not need to update the status here to approved?

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/PublicationService.java [hunk 4]
```diff
@@ -50,4 +50,35 @@ public String sendNotificationEmail(String emailAddress, String forename, String
...
         }
     }
+
+    public String sendNotificationEmailForSetupMediaAccount(String emailAddress, String fullName) {
+        JSONObject jsonObject = new JSONObject();
+        jsonObject.put("email", emailAddress);
+        jsonObject.put("isExisting", false);
+        jsonObject.put("fullName", fullName);
+        try {
+            return webClient.post().uri(new URI(url + "/notify/welcome-email"))
```
Comment (kurt-uk) [inline src/main/java/uk/gov/hmcts/reform/pip/account/management/service/PublicationService.java:60]: dont need new URI

**Review by kurt-uk — CHANGES_REQUESTED** [review]
youre also missing some updated integration tests which would have caught an issue in your account service

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java (hunk 2 above)
Comment (junaidiqbalmoj) [reply src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java:144]: It is working, I tested it.

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/PublicationService.java (hunk 4 above)
Comment (junaidiqbalmoj) [reply src/main/java/uk/gov/hmcts/reform/pip/account/management/service/PublicationService.java]: removed.

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java (hunk 1 above)
Comment (junaidiqbalmoj) [reply src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java:151]: I discussed with Chris S. This is sending email and it is not validating anything. So, keep it that way.

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java (hunk 2 above)
Comment (junaidiqbalmoj) [reply src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java]: got your point.

File: src/test/java/uk/gov/hmcts/reform/pip/account/management/service/AccountServiceTest.java [hunk 5]
```diff
@@ -234,6 +249,22 @@ void testAddUsers() {
...
+
+        Map<CreationEnum, List<?>> expected = new ConcurrentHashMap<>();
+        expected.put(CreationEnum.CREATED_ACCOUNTS, List.of());
+        expected.put(CreationEnum.ERRORED_ACCOUNTS, List.of());
+
+        PiUser user = new PiUser(UUID.randomUUID(), UserProvenances.PI_AAD, ID, EMAIL, Roles.INTERNAL_ADMIN_CTSC);
+        when(validator.validate(user)).thenReturn(Set.of());
+
+        assertEquals(expected, accountService.addUsers(List.of(user), EMAIL), "Returned maps should match");
+    }
```
[the remark this answers is not in the document]
Comment (junaidiqbalmoj) [reply src/test/java/uk/gov/hmcts/reform/pip/account/management/service/AccountServiceTest.java:266]: add another test to cover email scenario

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]
Some conflicts need fixing on this one

**Review by joshblackmoor — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java [hunk 6]
```diff
@@ -148,9 +154,30 @@ public Map<CreationEnum, List<?>> addUsers(List<PiUser> users, String issuerEmai
                 continue;
             }
 
+            Optional<MediaApplication> application =
+                mediaApplicationRepository.findByEmail(user.getEmail());
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java:158]: Rather than the application repository, should this really be looking at the pi user table or the azure active directory? he user table get's cleared out each month, so I'm thinking it won't be the right source of info

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

File: src/integrationTest/java/uk/gov/hmcts/reform/pip/account/management/controllers/AccountTest.java [hunk 7]
```diff
@@ -532,6 +562,41 @@ void testCreateMultipleSuccessUsers() throws Exception {
...
 
+        assertEquals(1, mappedResponse.get(CreationEnum.CREATED_ACCOUNTS).size(), "1 Users should be created");
+    }
+
+    @Test
+    void testCreateMultipleSuccessUsersWithDifferentEmails() throws Exception {
+
+        when(graphClient.users(any())).thenReturn(userRequestBuilder);
+        when(userRequestBuilder.buildRequest()).thenReturn(userRequest);
+        when(userRequest.get()).thenThrow(graphServiceException);
```
Comment (kurt-uk) [inline src/integrationTest/java/uk/gov/hmcts/reform/pip/account/management/controllers/AccountTest.java:573]: do you need to throw here? this looks like youre testing success?

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java [hunk 8]
```diff
@@ -57,6 +59,9 @@ public class AccountService {
     @Autowired
     UserRepository userRepository;
 
+    @Autowired
+    MediaApplicationRepository mediaApplicationRepository;
+
```
Comment (kurt-uk) [inline src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java:64]: doesnt look like you need this now

File: src/integrationTest/java/uk/gov/hmcts/reform/pip/account/management/controllers/AccountTest.java [hunk 9]
```diff
@@ -532,6 +562,41 @@ void testCreateMultipleSuccessUsers() throws Exception {
...
+                .post(PI_URL)
+                .content(objectMapper.writeValueAsString(List.of(validUser1, validUser2)))
+                .header(ISSUER_HEADER, ISSUER_EMAIL)
+                .contentType(MediaType.APPLICATION_JSON);
+
+        MvcResult response = mockMvc.perform(mockHttpServletRequestBuilder).andExpect(status().isCreated()).andReturn();
+        ConcurrentHashMap<CreationEnum, List<Object>> mappedResponse =
+            objectMapper.readValue(response.getResponse().getContentAsString(),
+                                   new TypeReference<>() {});
+
```
Comment (kurt-uk) [inline src/integrationTest/java/uk/gov/hmcts/reform/pip/account/management/controllers/AccountTest.java:599]: looking at account service code complexity i think it would be best to include int test that tests adding multiple users some valid and some duplicate

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/PublicationService.java [hunk 10]
```diff
@@ -40,11 +41,43 @@ public class PublicationService {
...
+            log.info(webClient.post().uri(url + WELCOME_EMAIL_URL)
+                .body(BodyInserters.fromValue(jsonObject)).retrieve()
+                .bodyToMono(String.class).block());
+            return true;
+
+        } catch (WebClientException ex) {
+            log.error(String.format("Request failed with error message: %s", ex.getMessage()));
+            return false;
+        }
+    }
```
Comment (kurt-uk) [inline src/main/java/uk/gov/hmcts/reform/pip/account/management/service/PublicationService.java:73]: please remember to check code that already exists, this code is directly under the code you added, the only thing you needed to add was the full name field, please remove and add to the body of the method that already exists

File: src/test/java/uk/gov/hmcts/reform/pip/account/management/service/AccountServiceTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kurt-uk) [reply src/test/java/uk/gov/hmcts/reform/pip/account/management/service/AccountServiceTest.java:291]: looking at the code you added in last commit, youve added at least 6 possible outcomes with the if statement complexities and try catch's, however i can only see 2 tests added, youll still need to make sure youre covering all the code you write

**Review by kurt-uk — CHANGES_REQUESTED** [review]

File: src/integrationTest/java/uk/gov/hmcts/reform/pip/account/management/controllers/AccountTest.java (hunk 7 above)
Comment (junaidiqbalmoj) [reply src/integrationTest/java/uk/gov/hmcts/reform/pip/account/management/controllers/AccountTest.java]: My mistake, it should return user object not exception.

File: src/integrationTest/java/uk/gov/hmcts/reform/pip/account/management/controllers/AccountTest.java (hunk 9 above)
Comment (junaidiqbalmoj) [reply src/integrationTest/java/uk/gov/hmcts/reform/pip/account/management/controllers/AccountTest.java:608]: testCreateMultipleSuccessUsers (This test is trying to create two users with same email address, but it creates only one because second is categorised as duplicate).
testCreateMultipleSuccessUsersWithDifferentEmails (this test creates two different users having different email address)

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/PublicationService.java (hunk 10 above)
Comment (junaidiqbalmoj) [reply src/main/java/uk/gov/hmcts/reform/pip/account/management/service/PublicationService.java]: Done

**Review by junaidiqbalmoj — COMMENTED** [review]
add more unit tests

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/PublicationService.java (hunk 10 above)
Comment (kurt-uk) [reply src/main/java/uk/gov/hmcts/reform/pip/account/management/service/PublicationService.java]: im still seeing it, you need to delete the method `sendMediaNotificationEmail` as it already exists as the first method in this class, delete that enture method and use the one that already exists called `sendNotificationEmail`

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java [hunk 11]
```diff
@@ -148,10 +149,36 @@ public Map<CreationEnum, List<?>> addUsers(List<PiUser> users, String issuerEmai
                 continue;
             }
 
-            PiUser addedUser = userRepository.save(user);
-            createdAccounts.add(addedUser.getUserId());
+            try {
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java:152]: I havn't tried it, but if the user already exists in Azure (which the frontend will try to create first when you click 'Submit'), and therefore that request fails, will it get to this part of the code?

If it does then all fine :) However if not we have an issue where it will fail at that bit then never reach this bit of code

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java (hunk 11 above)
Comment (junaidiqbalmoj) [reply src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java:152]: This section of code adding user in pi_user table. addAzureAccounts that's the function which is trying to add user into  azure AAD.

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java [hunk 12]
```diff
@@ -148,10 +149,36 @@ public Map<CreationEnum, List<?>> addUsers(List<PiUser> users, String issuerEmai
...
+                User userAzure = azureUserService.getUser(user.getEmail());
+
+                if (userRepository.findByEmail(user.getEmail()).isPresent()
```
[the remark this answers is not in the document]
Comment (junaidiqbalmoj) [reply src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java]: updated to code to use newly created repo function findByEmailAndProvenanceId

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java (hunk 11 above)
Comment (kurt-uk) [reply src/main/java/uk/gov/hmcts/reform/pip/account/management/service/AccountService.java:152]: @ChrisS1512 the FE will use the response of the first request to fire the second request so it should not get to this part of the code, IIRC, FE takes the accounts from CREATED_ACCOUNTS to build the 2nd request to add to PI db, so if 1st request failed and successfully adds them to ERRORED_ACCOUNTS you shouldnt get here

File: src/integrationTest/java/uk/gov/hmcts/reform/pip/account/management/controllers/AccountTest.java (hunk 9 above)
Comment (kurt-uk) [reply src/integrationTest/java/uk/gov/hmcts/reform/pip/account/management/controllers/AccountTest.java:608]: i cant see the outcome you described in those tests, all the tests only test for 1 outcome, errored account,
Screenshot 2022-07-04 at 09 38 04
i can only see this outcome in the tests youve added, the created accounts is an empty list,

File: src/main/java/uk/gov/hmcts/reform/pip/account/management/database/UserRepository.java [hunk 13]
```diff
@@ -17,4 +17,9 @@ List<PiUser> findExistingByProvenanceId(@Param("provUserId") String provenanceUs
                                             @Param("userProv") String userProvenance);
 
     Optional<PiUser> findByUserId(UUID userId);
+
+    @Query(value = "SELECT * FROM pi_user WHERE email=:email AND user_provenance=:userProv",
+        nativeQuery = true)
+    Optional<PiUser> findByEmailAndProvenanceId(@Param("email") String provenanceUserId,
```
Comment (kurt-uk) [inline src/main/java/uk/gov/hmcts/reform/pip/account/management/database/UserRepository.java:23]: not really provenenaceUserId, its email, those are 2 very different values, careful not to cause confusion doing this. would name appropiately

**Review by kurt-uk — APPROVED** [review]
approving to not block, however all comments still stand, will leave to @ChrisS1512 to police it

**Review by ChrisS1512 — APPROVED** [review]
